libavfilter/sf_text2bitmap.c

```diff
@@ -0,0 +1,539 @@
+    for (ASS_Image *i = image; i != NULL; i = i->next, area->nb_colors++);
+    set_area_bounds(image, area);
+
+    area->pal[0] = 0;                                   // Set transparent
+    area->buf[0] = av_buffer_allocz(area->w * area->h * sizeof(uint8_t));   // Init to transparent
+    for (int color = 1; image != NULL; image = image->next, color++)
+    {
+        area->pal[color] = TO_AREA_COLOR(image->color);   // Set color
+        for (int y = 0; y < image->h; y++)                  // Set pixels
+            for (int x = 0; x < image->w; x++)              //
+                if (image->bitmap[y * image->stride + x] >= ALPHA_THRESHOLD)
+                {
+                    int x_area = image->dst_x + x - area->x;
+                    int y_area = image->dst_y + y - area->y;
+                    area->buf[0]->data[y_area * area->w + x_area] = color;
+                }
+    }
+}
+*/
+#define ALPHA_MAX ((1 << (sizeof(uint8_t) << 3)) - 1)
+static void palettize_image(PalettizeContext *const s,
+                            const int w, const int h,
+                            uint8_t *src_data, const int src_linesize,
+                            uint8_t *dst_data, const int dst_linesize,
+                            uint32_t *dst_pal, const int num_output_colors);
+
+static void ass_image_to_area_palletization(ASS_Context *context, ASS_Image *image, AVSubtitleArea *area)
+{
+    unsigned nb_pixels;
+    size_t image_rgba_size;
+    uint8_t *image_rgba;
+
+    set_area_bounds(image, area);
+
+    // Create rgba image
+    nb_pixels = area->w * area->h;
+    image_rgba_size = nb_pixels * 4 * sizeof(uint8_t);
+    image_rgba = (uint8_t *)malloc(image_rgba_size);
+    memset(image_rgba, 0, image_rgba_size);        // Set all to transparent
+    for (; image != NULL; image = image->next)
+    {
+        // Set pixels
+        for (int y = 0; y < image->h; y++)
+            for (int x = 0; x < image->w; x++) {
+                int x_area = image->dst_x + x - area->x;
+                int y_area = image->dst_y + y - area->y;
+                uint8_t alpha_b = image_rgba[4 * (y_area * area->w + x_area) + 3];
+                uint8_t alpha_a = AA(image->color) * image->bitmap[y * image->stride + x] >> 8;
+                uint8_t alpha = (ALPHA_MAX * alpha_a + alpha_b * (ALPHA_MAX - alpha_a)) / ALPHA_MAX;
+                image_rgba[4 * (y_area * area->w + x_area) + 3] = alpha;
+                if (alpha == 0) continue;       // The components don't matter (plus we get FPE otherwise)
+                for (int c = 0; c < 3; c++) {
+                    uint8_t comp_b = image_rgba[4 * (y_area * area->w + x_area) + c];
+                    uint8_t comp_a = (image->color >> ((c + 1) * 8)) & 0xFF;
+                    uint8_t comp = (comp_a * alpha_a * ALPHA_MAX + comp_b * alpha_b * (ALPHA_MAX - alpha_a)) / (alpha * ALPHA_MAX);
+                    image_rgba[4 * (y_area * area->w + x_area) + c] = comp;
+                }
+            }
+    }
```

Comment:
Why don't you use ff_draw_ like the other implementations?

Reply:
First time I hear of it, thanks!

Reply:
> First time I hear of it, thanks!

First time?? - What I mean is how vf_subtitles and vf_overlaytextsubs are doing it..

Don't forget the ff_draw_init() call for transparency handling (like textsub2video does)